File: Java/LeetCode/lc64/Solution.java
package lc64;
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        int[][] dp = new int[m][n];
        dp[0][0] = grid[0][0];

        for(int i = 1; i < m; i ++){
            dp[i][0] = dp[i-1][0]+grid[i][0];
        }
        for(int i = 1; i < n; i ++){
            dp[0][i] = dp[0][i-1]+grid[0][i];
        }

        for(int col = 1; col < m; col++){
            for(int row = 1; row < n; row++){
                dp[col][row] = Math.min(dp[col-1][row],dp[col][row-1])+grid[col][row];
            }
        }

        return dp[m-1][n-1];
    }
}

File: Rust/LeetCode/lc64/solution.rs
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let mut dp = vec![0; grid[0].len()+1];

        for i in 1..dp.len(){
            dp[i] = dp[i-1]+grid[0][i-1];
        }

        dp[0] = i32::MAX;
        for row in 1..grid.len(){
            for col in 1..=grid[0].len(){
                dp[col] = dp[col].min(dp[col-1])+grid[row][col-1];    
            }
        }

        dp[grid[0].len()]
    }
}